### PR TITLE
S3 parameters endpoint

### PR DESCRIPTION
**Pull request by goto-bus-stop**
This PR adds a route where uppy clients can fetch signed upload parameters to use to upload files straight to an S3 bucket.

I added it as a new controller file, since it's not exactly a provider like google drive or dropbox are (no list/auth/etc methods).

<strike>I mushed this together based on gists and stackoverflow answers. I think it's possible to use the AWS SDK for this instead though, which would be nicer. I'll try that soon.

So this isn't really ready for an in-depth review, but comments on if just adding a controller like this is OK or if it should go somewhere else would be nice.</strike>
Now uses the AWS SDK.

### CODE REVIEW

File: src/server/controllers/s3.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  key: process.env.UPPYSERVER_AWS_KEY,
+  secret: process.env.UPPYSERVER_AWS_SECRET,
+  bucket: process.env.UPPYSERVER_AWS_BUCKET,
+  region: process.env.UPPYSERVER_AWS_REGION,
+  conditions: []
+}
+
+module.exports = function s3 (config) {
+  config = Object.assign({}, defaultConfig, config)
+  const client = new S3({
```
Comment (goto-bus-stop) [inline src/server/controllers/s3.js:12]: Hmm, this thing could perhaps be passed in by the user as well if they're using the `pluggable` way.

**Comment by ifedapoolarewaju** [comment]
Looking good to me so far 👍

I think considering that Transloadit and Instagram APIs also have support for signed requests, we might at some point want to have those in uppy-server as well. Hence, following the pattern of naming controller files after actions such as `list`, `get`, `auth`, `logout`, `sign-request` might make sense.

for now we could at least ensure that the routes follow the pattern `:providerName/:action`, i.e `s3/sign-request` or `s3/sign`. This way if we decide to refactor into the pattern above ☝️, the routes wouldn't have to change.

File: src/server/controllers/s3.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+const router = require('express').Router
+const ms = require('ms')
+const S3 = require('aws-sdk/clients/s3')
+
+const defaultConfig = {
+  acl: 'public-read',
+  key: process.env.UPPYSERVER_AWS_KEY,
```
Comment (ifedapoolarewaju) [inline src/server/controllers/s3.js]: maybe we should pass the default env variables via the [stand-alone server](https://github.com/transloadit/uppy-server/blob/master/src/standalone.js#L56) and not have the pluggable app worry about environment variables.

Comment (goto-bus-stop) [reply src/server/controllers/s3.js]: 👍 good suggestion, thanks!

**Comment by goto-bus-stop** [comment]
> I think considering that Transloadit and Instagram APIs also have support for signed requests, we might at some point want to have those in uppy-server as well. Hence, following the pattern of naming controller files after actions such as `list`, `get`, `auth`, `logout`, `sign-request` might make sense.

`sign` was a bit of a misnomer, the route actually does a lot more. It generates upload parameters (an url to upload to and form fields), and it just happens to also include a signature. I renamed it to `/s3/params` now to be a bit more clear about that 😛 We'll also probably have more s3-specific routes in the future, eg to start/sign/complete multipart uploads.

> for now we could at least ensure that the routes follow the pattern `:providerName/:action`, i.e `s3/sign-request` or `s3/sign`. This way if we decide to refactor into the pattern above ☝️, the routes wouldn't have to change.

Aye, done 👍

**Comment by goto-bus-stop** [comment]
I don't understand the test failure, does it look familiar to you @ifedapoolarewaju?

**Comment by ifedapoolarewaju** [comment]
@goto-bus-stop this is new to me, but look [what I found](https://stackoverflow.com/questions/43837596/projects-map-is-not-a-function-for-jest-cli)

I did update jest-cli, but that wasn't on this branch. It was on master during the hackathon.

so maybe changing [this](https://github.com/transloadit/uppy-server/blob/master/package.json#L83) to `"test": "npm run lint && env UPPYSERVER_DATADIR=./test/output ./node_modules/.bin/jest",` should fix it?

**Comment by goto-bus-stop** [comment]
Ah, thanks!
These do have different versions now actually: https://github.com/transloadit/uppy-server/blob/master/package.json#L23-L24 will try syncing them first

**Comment by goto-bus-stop** [comment]
That did it, rebased & updated the npm and yarn lockfiles :)

I just want to make using s3 optional, since we won't be running that on the transloadit cloud (right?), but it's only for users who are running their own uppy-server. Other than that I think this is ready.

**Comment by ifedapoolarewaju** [comment]
LGTM! 👍

**Pull request merged**